### ISSUE TITLE
Poor use of MPS allocation protocol

### ISSUE THREAD

**Issue by rptb1** [dylan-lang/opendylan#235]
This is an issue that has been around since the early days of DylanWorks, and was going to be fixed "soon" forever. It could possibly improve overall performance a great deal.

Currently, Open Dylan allocates objects by calling a function that initializes the object type and fills the rest with zeros.  The compiled code then fills the object in again with the initial slot values.  Objects are initialized twice.

The MPS has a non-locking allocation protocol that can be inlined into just a few instructions and avoids the need to fill the object.  The compiler could be adjusted to output inline allocations using this protocol.

In straightlined C-like pseudocode, the current sequence on allocation is roughly:

``` C
do {
  p = reserve(ap, size)
  fill_with_zero(p, size);
  p[0] = object type
} while(!commit(ap, size));
...
p[1] = slot value A
...
p[2] = slot value B
...
```

In in the current code, neither reserve nor commit are inlined.  The intended way to use the protocol is:

``` C
...
temp[0] = slot value A
...
temp[1] = slot value B
...
do {
  p = reserve(ap, size);
  p[0] = object type
  p[1] = temp[0]
  p[2] = temp[1]
} while(!commit(ap, size));
```

Note that in fact you can avoid using temps if the code to calculate the slot values does not allocate.  The only restriction is you can't nest allocations on the same allocation point (AP). Dylan would probably want to avoid any code with side-effects, as it may be repeated if commit fails.

A word about the loop.  Commit can return false if there is a GC "flip" between the reserve and commit.  In that case, the MPS might have moved all the objects in the world, and the half-initialized object is invalid.  This is a very unlikely, as long as the time between reserve and commit is short.  The purpose of commit is to declare to the MPS that the object is now initialized and can be scanned by the Dylan format methods.

The reserve and commit operations are designed to be inlined in just a few instructions.  They are defined in macros in mps.h, but they look like roughly this (simplified very slightly):

Reserve

``` C
if (ap->alloc + size < ap->limit) {
  ap->alloc += size;
  p = ap->init;
} else
  p = mps_ap_fill();
```

Commit

``` C
ap->init = ap->alloc;
if (ap->limit == 0)
  mps_ap_trip();
```

In other words, reserve is just a pointer bump with an overflow check, and commit is an assignment and a flag check.  The ap pointer should generally be in a global register or nearby in thread local storage.

I would be very happy to discuss further details here and will help anyone who takes on the task of improving the allocation protocol. It's a shame that Open Dylan isn't using the fast protocol I designed for it!

See also http://colabti.org/irclogger/irclogger_log/dylan?date=2012-08-08 for #dylan discussion on this topic from yesterday.
